crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs

```diff
@@ -42,57 +45,119 @@ impl Ssa {
+        let instruction = &dfg[instruction_id];
+        match instruction {
+            Instruction::Binary(binary) => {
+                let result_acir_var = self.convert_ssa_binary(binary, dfg);
+                let result_value_id = dfg
+                    .instruction_results(instruction_id)
+                    .first()
+                    .expect("Binary ops have a single result");
+                self.ssa_value_to_acir_var.insert(*result_value_id, result_acir_var);
+            }
+            _ => todo!(),
+        }
     }
 
-    /// Returns all of the instructions in the program.
-    ///
-    /// This method assumes that all functions and basic blocks have
-    /// been inlined and so we only need to look in the entry block,
-    /// in the first function for all instructions.
-    fn instructions(ssa: &Ssa) -> Vec<InstructionId> {
-        assert_eq!(
-            ssa.functions.len(),
-            1,
-            "expected only a single function to be present with all other functions being inlined."
-        );
-        let (_, entry_function) = ssa.functions.iter().next().expect(
-            "This line should be infallible, since we checked that there is exactly one function",
-        );
+    fn convert_ssa_return(&mut self, terminator: &TerminatorInstruction, dfg: &DataFlowGraph) {
+        let return_values = match terminator {
+            TerminatorInstruction::Return { return_values } => return_values,
+            _ => unreachable!("ICE: Program must have a singular return"),
+        };
+        for value_id in return_values {
+            let acir_var = self
+                .ssa_value_to_acir_var
+                .get(value_id)
+                .expect("ICE: Return of value not yet encountered");
+            self.acir_context.return_var(*acir_var);
+        }
+    }
 
-        let entry_block_id = entry_function.entry_block();
-        let dfg = &entry_function.dfg;
-        let entry_block = &dfg[entry_block_id];
+    /// Converts an SSA value into a `AcirVar`
+    fn convert_ssa_value(&mut self, value_id: ValueId, dfg: &DataFlowGraph) -> AcirVar {
+        let value = &dfg[value_id];
+        if let Some(acir_var) = self.ssa_value_to_acir_var.get(&value_id) {
+            return *acir_var;
+        }
+        let acir_var = match value {
+            Value::NumericConstant { constant, .. } => {
+                let field_element = &dfg[*constant].value();
+                self.acir_context.add_constant(*field_element)
+            }
+            Value::Intrinsic(..) => todo!(),
+            Value::Function(..) => unreachable!("ICE: All functions should have been inlined"),
+            // `ssa_value_to_acir_var.get(value_id)` should have already returned the
+            // corresponding `AcirVar`.
+            Value::Instruction { .. } => {
+                unreachable!("ICE: Instruction result value used before instruction encountered")
```

Comment:
> Instruction result value used before instruction encountered

Can you clarify what this means? If a method is making assumptions, lets put it as part of the method documentation like what I did for `assumptions`, with that the panic messages can be shorter

Reply:
function comment elborated